File: 3601-find-the-k-th-character-in-string-game-ii/find-the-k-th-character-in-string-game-ii.cpp
class Solution {
public:
    char kthCharacter(long long k, vector<int>& operations) {
        int n = operations.size();
        long long val = k;
        int shifts = 0;

        while (val > 1) {
            int jumps = std::floor(log2(val - 1));
            val = val - std::pow(2, jumps);
            shifts += operations[jumps];
        }

        return 'a' + (shifts % 26);
    }
};

File: 3601-find-the-k-th-character-in-string-game-ii/find-the-k-th-character-in-string-game-ii.java
class Solution {
    public char kthCharacter(long k, int[] operations) {
        int n = operations.length;
        long val = k;
        int shifts = 0;

        while (val > 1) {
            int jumps = (int)Math.ceil(Math.log(val) / Math.log(2));
            val = val - (long)Math.pow(2, jumps - 1);
            shifts += operations[jumps - 1];
        }

        return (char)('a' + (int)(shifts % 26));
    }
}